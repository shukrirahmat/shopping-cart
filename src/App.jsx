import { Link, useParams } from "react-router-dom";
import Home from "./components/Home";
import Shop from "./components/Shop";
import Cart from "./components/Cart";
import styles from "./App.module.css";
import "./index.css";
import { useEffect, useState } from "react";
import cartIcon from "./assets/icons/cart-outline.svg";

function useItemsData() {
  const [items, setItems] = useState([]);
  const [error, setError] = useState(null);
  const [loading, setLoading] = useState(true);

  useEffect(() => {
    fetch("https://fakestoreapi.com/products", { mode: "cors" })
      .then((response) => {
        if (response.status >= 400) {
          throw new Error("server error");
        }
        return response.json();
      })
      .then((response) => {
        //console.log(response);
        setItems(response);
      })
      .catch((error) => setError(error))
      .finally(() => setLoading(false));
  }, []);

  return { items, error, loading };
}

function App() {
  const { name } = useParams();
  const { items, error, loading } = useItemsData();
  const [cartItems, setCartItems] = useState([]);

  function addToCart(item, amount) {
    let newList = cartItems.slice();
    let itemExist = false;

    newList = newList.map((e) => {
      if (e.id === item.id) {
        const newamount = e.amount + amount;
        const newItem = { ...e, amount: newamount};
        itemExist = true;
        return newItem;
      } else {
        return e
      }
    });

    if (!itemExist) {
      const newItem = { ...item, amount: amount };
      newList.push(newItem);
    }
    setCartItems(newList);
  }

  function removeItem(id) {
    const newList = cartItems.filter((item) => item.id !== id);
    setCartItems(newList);
  }

  function modifyAmount(id, newAmount) {
    const newList = cartItems.map((e) => {
      if (e.id === id) {
        const newItem = { ...e, amount: newAmount};
        return newItem;
      } else {
        return e
      }
    });
    setCartItems(newList);
  }

  if (error)
    return <p className={styles.errortext}>A network error was encountered.</p>;

  return (
    <div className={styles.background}>
      <div className={styles.header}>
        <div className={styles.headWrapper}>
          <p className={styles.logo}>Yangtze</p>
          <nav className={styles.linksWrapper}>
            <Link to="/" className={styles.link}>
              HOME
            </Link>
            <Link to="/shop" className={styles.link}>
              SHOP
            </Link>
            <Link to="/cart" className={styles.link}>
              <div className={styles.cartIconWrapper}>
                <img className={styles.cartIcon} src={cartIcon}></img>
                {cartItems.length === 0 ? null : (
                  <p className={styles.cartTotal}>{cartItems.length}</p>
                )}
              </div>
            </Link>
          </nav>
        </div>
      </div>
      <div>
        {name === "shop" ? (
          <Shop items={items} loading={loading} addToCart={addToCart} />
        ) : name === "cart" ? (
          <Cart cartItems={cartItems} removeItem={removeItem} modifyAmount={modifyAmount}/>
        ) : (
          <Home />
        )}
      </div>
    </div>
  );
}

export default App;
