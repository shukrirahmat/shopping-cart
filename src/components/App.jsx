import { Link, useParams } from "react-router-dom";
import Home from "./Home";
import Shop from "./Shop";
import Cart from "./Cart";
import styles from "../styles/App.module.css";
import { useEffect, useState } from "react";
import cartIcon from "../assets/icons/cart-outline.svg";
import "../index.css";
import PropTypes from "prop-types";

function App({items, error, loading, initialPath}) {

  let { name } = useParams();
  if (initialPath) {
    name = initialPath;
  }

  const [cartItems, setCartItems] = useState([]);

  function addToCart(item, amount) {
    let newList = cartItems.slice();
    let itemExist = false;

    newList = newList.map((e) => {
      if (e.id === item.id) {
        const newamount = e.amount + amount;
        const newItem = { ...e, amount: newamount};
        itemExist = true;
        return newItem;
      } else {
        return e
      }
    });

    if (!itemExist) {
      const newItem = { ...item, amount: amount };
      newList.push(newItem);
    }
    setCartItems(newList);
  }

  function removeItem(id) {
    const newList = cartItems.filter((item) => item.id !== id);
    setCartItems(newList);
  }

  function modifyAmount(id, newAmount) {
    const newList = cartItems.map((e) => {
      if (e.id === id) {
        const newItem = { ...e, amount: newAmount};
        return newItem;
      } else {
        return e
      }
    });
    setCartItems(newList);
  }

  function countAllItems() {
    let total = 0;
    cartItems.forEach(i => {
      total = total + i.amount;
    })

    return total;
  }

  if (error)
    return <p className={styles.errortext}>A network error was encountered.</p>;

  return (
    <div className={styles.background}>
      <div className={styles.header}>
        <div className={styles.headWrapper}>
          <p className={styles.logo}>Yangtze</p>
          <nav className={styles.linksWrapper}>
            <Link to="/" className={styles.link}>
              HOME
            </Link>
            <Link to="/shop" className={styles.link}>
              SHOP
            </Link>
            <Link to="/cart" className={styles.link}>
              <div className={styles.cartIconWrapper}>
                <img className={styles.cartIcon} src={cartIcon}></img>
                {cartItems.length === 0 ? null : (
                  <p role="cartNum" className={styles.cartTotal}>{countAllItems()}</p>
                )}
              </div>
            </Link>
          </nav>
        </div>
      </div>
      <div>
        {name === "shop" ? (
          <Shop items={items} loading={loading} addToCart={addToCart} />
        ) : name === "cart" ? (
          <Cart cartItems={cartItems} removeItem={removeItem} modifyAmount={modifyAmount}/>
        ) : (
          <Home />
        )}
      </div>
    </div>
  );
}

App.propTypes = {
  items: PropTypes.array,
  error: PropTypes.object,
  loading: PropTypes.bool,
  initialPath: PropTypes.string,
};

export default App;
