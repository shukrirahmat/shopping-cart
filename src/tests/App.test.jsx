import { describe, it, expect } from "vitest";
import { render , screen} from "@testing-library/react";
import { BrowserRouter } from "react-router-dom";
import App from "../components/App";
import userEvent from "@testing-library/user-event";

describe("App component", () => {
    const item = {
        id: '69',
        title: 'test product',
        price: 13.5,
        description: 'lorem ipsum set',
        image: 'https://i.pravatar.cc',
        category: 'electronic'
    }

    it("renders App page", () => {
        const {container} = render(<BrowserRouter><App items={[item]} error={null} loading={false}/></BrowserRouter>)
        expect(container).toMatchSnapshot();
    })
})