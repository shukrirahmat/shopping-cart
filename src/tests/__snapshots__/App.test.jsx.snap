// Vitest Snapshot v1, https://vitest.dev/guide/snapshot.html

exports[`App component > renders App page 1`] = `
<div>
  <div
    class="_background_61351a"
  >
    <div
      class="_header_61351a"
    >
      <div
        class="_headWrapper_61351a"
      >
        <p
          class="_logo_61351a"
        >
          Yangtze
        </p>
        <nav
          class="_linksWrapper_61351a"
        >
          <a
            class="_link_61351a"
            href="/"
          >
            HOME
          </a>
          <a
            class="_link_61351a"
            href="/shop"
          >
            SHOP
          </a>
          <a
            class="_link_61351a"
            href="/cart"
          >
            <div
              class="_cartIconWrapper_61351a"
            >
              <img
                class="_cartIcon_61351a"
                src="/src/assets/icons/cart-outline.svg"
              />
            </div>
          </a>
        </nav>
      </div>
    </div>
    <div
      class="_bottom_61351a"
    >
      <div
        class="_wrapper_b140de"
      >
        <img
          alt="Store Photo"
          class="_photo_b140de"
          src="/src/assets/images/storephoto.jpg"
        />
        <div
          class="_welcomeText_b140de"
        >
          <p
            class="_title_b140de"
          >
            <b>
              Shops online with fast and reliable delivery
            </b>
          </p>
          <p>
            Made from top-tier materials, check out of our high quality products! 
          </p>
          <a
            class="_shopBtn_b140de"
            href="/shop"
          >
            Shop now
          </a>
        </div>
      </div>
    </div>
  </div>
</div>
`;
